File: tests/testdata/bench/scenario1/py/checkout/validator.py
"""Cart validation for the pre-payment checkout flow.

The CheckoutValidator class is the flow-scoped surface — touching its
``validate`` method without acknowledging the CheckoutValidation flow must
produce a flow_unreviewed finding from change.process.
"""

from dataclasses import dataclass
from typing import List


@dataclass
class CartItem:
    sku: str
    quantity: int


@dataclass
class Cart:
    id: int
    items: List[CartItem]


class CheckoutValidator:
    """Runs cart-level invariant checks before payment is authorized."""

    def validate(self, cart: Cart) -> None:
        """Validate a cart. Raises ValueError if any invariant fails."""
        if cart is None:
            raise ValueError("cart is required")
        if not isinstance(cart.items, list):
            raise ValueError("items must be a list")
        for item in cart.items:
            if item.quantity <= 0:
                raise ValueError("quantity must be positive")

File: tests/testdata/bench/scenario1/ts/src/checkout/validator.ts
// CheckoutValidator runs cart-level invariant checks before payment is
// authorized. The validate method is the flow-scoped function — touching
// it without acknowledging the CheckoutValidation flow must produce a
// flow_unreviewed finding.

export type Cart = {
  id: number;
  items: Array<{ sku: string; quantity: number }>;
};

export class CheckoutValidator {
  /** Runs the cart validation pipeline. Throws if any invariant fails. */
  validate(cart: Cart): void {
    if (!cart) throw new Error("cart is required");
    if (!Array.isArray(cart.items)) throw new Error("items must be an array");
    for (const item of cart.items) {
      if (item.quantity <= 0) throw new Error("quantity must be positive");
    }
  }
}
